src/frontend/src/optimizer/plan_node/logical_multi_join.rs

```diff
@@ -111,6 +115,269 @@ impl PlanTreeNode for LogicalMultiJoin {
     }
 }
 
+#[derive(Debug)]
+struct ConnectedComponentLabeller {
+    vertex_to_label: HashMap<usize, usize>,
+    labels_to_vertices: HashMap<usize, HashSet<usize>>,
+    labels_to_edges: HashMap<usize, HashSet<(usize, usize)>>,
+}
+
+impl ConnectedComponentLabeller {
+    fn new(vertices: usize) -> Self {
+        let mut vertex_to_label = HashMap::with_capacity(vertices);
+        let mut labels_to_vertices = HashMap::with_capacity(vertices);
+        let labels_to_edges = HashMap::new();
+        for i in 0..vertices {
+            vertex_to_label.insert(i, i);
+            labels_to_vertices.insert(i, vec![i].into_iter().collect());
+        }
+        Self {
+            vertex_to_label,
+            labels_to_vertices,
+            labels_to_edges,
+        }
+    }
+
+    fn add_edge(&mut self, v1: usize, v2: usize) {
+        let v1_label = *self.vertex_to_label.get(&v1).unwrap();
+        let v2_label = *self.vertex_to_label.get(&v2).unwrap();
+
+        let (new_label, old_label) = if v1_label < v2_label {
+            (v1_label, v2_label)
+        } else {
+            // v1_label > v2_label
+            (v2_label, v1_label)
+        };
+
+        {
+            let edges = self
+                .labels_to_edges
+                .entry(new_label)
+                .or_insert_with(HashSet::new);
+
+            let new_edge = if v1 < v2 { (v1, v2) } else { (v2, v1) };
+            edges.insert(new_edge);
+        }
+
+        if v1_label == v2_label {
+            return;
+        }
+
+        // Reassign to the smaller label
+        let old_vertices = self.labels_to_vertices.remove(&old_label).unwrap();
+        self.labels_to_vertices
+            .get_mut(&new_label)
+            .unwrap()
+            .extend(old_vertices.iter());
+        for v in old_vertices {
```

Comment:
Seems `Union-Find` can help find the connected components more efficiently. 

But the number of joins won't be too many, the current implementation should also be ok.

Reply:
yeah its definitely not best performance. Didn't want to write a custom data structure just for it...